src/selectors/lesson.js

```diff
@@ -1,76 +1,124 @@
-      const lesson = filteredLessons[lessonPath];
-      return lesson.indexed ? {...res, [lessonPath]: lesson} : res;
-    }, {});
+export const getFilteredLessonLevelsInCourse = createCachedSelector(
+  // Input selectors:
+  getFilteredLessonsInCourse,
+  (state, course) => course,
+
+  // Output selector (resultfunc):
+  (lessons, course) => {
+    const levels = new Set;
+    for (const lesson of lessons) {
+      levels.add(getLessonLevel(course, lesson));
+    }
+    return [...levels].sort();
   }
+)(
+  (state, course) => course
 );
 
 /**
- * Creates an object {<level>: [lessonA, lessonB, ...], ...}
- * where the lessons available given your current filter are sorted according to level
- * Input props: courseName (string, optional)
+ * Get filtered lessons for all courses.
+ * @param {object} state The redux state object.
+ * @returns {object} An object where the keys are the courses, and the values are arrays of courses, e.g.
+ * {
+ *   scratch: ['astrokatt', 'straffespark', ...]
+ *   python: ['bokstaver', 'fargespill', ...]
+ *   ...
+ * }
  */
-export const getLessonsByLevel = createSelector(
-  [getFilteredAndIndexedLessons],
-  (lessons = {}) => {
-    // Get lessons grouped by level
-    return Object.keys(lessons).reduce((res, lessonId) => {
-      const lesson = lessons[lessonId];
-      const level = lesson.level;
+export const getFilteredLessons = createSelector(
+  // Input selectors:
+  (state) => state.filter, // See structure in INITIAL_STATE in src/reducers/filter.js
 
-      // Ignore lessons without level
-      if (level != null) {
-        if (res.hasOwnProperty(level)) res[level].push(lesson);
-        else res[level] = [lesson];
-      }
+  // Output selector (resultfunc):
+  (filter) => {
+    const filteredLessons = {};
+    for (const course of getAllCourses()) {
+      // We call getFilteredLessonsInCourse directly (instead of having it
+      // as an input selector) because it needs the extra 'course' argument.
+      // But we don't want to supply getFilteredLessonsInCourse the whole state,
+      // since getFilteredLessons would need to get it through an input selector,
+      // and that would make getFilteredLessons recalculate every time anything in
+      // the state changes, not only when the filter changes.
+      // Instead, we recreate the redux state object with only 'filter' in it,
+      // since that is the only part of state that getFilteredLessonsInCourse needs.
+      // Note that if the implementation of getFilteredLessonsInCourse changes so that
+      // it uses other parts of the state, we will need to supply those, too.
+      //
+      // An alternative solution would be to separate out the resultfunc of
+      // getFilteredLessonsInCourse, and memoize it ourselves in a smart way,
+      // and then use this function in both getFilteredLessonsInCourse and getFilteredLessons.
+      // If so, we would not need to use reselect nor re-reselect for getFilteredLessonsInCourse.
```

Comment:
Do what is suggested in the lower part of the comment. Otherwise testing will be hard.
Using fast-memoize this should be pretty easy.

Reply:
Changed my mind, just simplify without optimizing for caching. It is fast enough anyway.